### PR TITLE
✨Fix up e2e conformance tests (release 0.4)

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:
- Use python script for envsubst
- Add / Delete keypair if not preset
- Avoid printing unnecessary stuff
- Support specifying a lookup org for images
- option for skipping creation of the AMI image

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: hack/heartbeat_account.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# limitations under the License.
+
+import sys
+import httplib
+import urllib
+import os
+import time
+
+BOSKOS_HOST=os.environ.get("BOSKOS_HOST", "boskos")
+BOSKOS_RESOURCE_NAME=os.environ['BOSKOS_RESOURCE_NAME']
```
Comment (detiber) [inline hack/heartbeat_account.py]: Is anything actually setting `BOSKOS_RESOURCE_NAME`?

Comment (detiber) [reply hack/heartbeat_account.py]: After some digging, it looks like this also affects the `hack/checkin_account.py` script, where I think we also have the same issue.

Comment (dims) [reply hack/heartbeat_account.py]: yes, it's easy to miss. we print the variable in the python script `checkout_account.py` and then source that stdout in:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/scripts/ci-conformance.sh#L52

Comment (dims) [reply hack/heartbeat_account.py]: then both checkout and heartbeat scripts get it from the env var.

File: examples/machinedeployment/machinedeployment.yaml [hunk 2]
```diff
@@ -6,7 +6,7 @@ metadata:
     cluster.x-k8s.io/cluster-name: ${CLUSTER_NAME}
     nodepool: nodepool-0
 spec:
-  replicas: 1
+  replicas: 2
```
Comment (aojea) [inline examples/machinedeployment/machinedeployment.yaml]: is this a cluster only with 2 nodes?

Comment (detiber) [reply examples/machinedeployment/machinedeployment.yaml]: 2 worker nodes (minimum required for conformance testing), there should also be 3 control plane machines as well.

Comment (dims) [reply examples/machinedeployment/machinedeployment.yaml]: 1 bastion node, 3 controllers, 2 workers (with this change included)

File: hack/heartbeat_account.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+USER = "cluster-api-provider-aws"
+
+if __name__ == "__main__":
+    count = 0
+    # keep sending heart beat for 3 hours
+    while count < 180:
+        conn = httplib.HTTPConnection(BOSKOS_HOST)
+        print "POST-ing heartbeat for resource %s to %s" % (BOSKOS_RESOURCE_NAME, BOSKOS_HOST)
+        conn.request("POST", "/update?%s" % urllib.urlencode({
```
Comment (aojea) [inline hack/heartbeat_account.py]: maybe use it inside a `try ... except` block to handle timeouts per example or network hiccups ??

File: scripts/ci-conformance.sh [hunk 4]
```diff
@@ -69,7 +83,8 @@ if grep -iqF "$(echo "${AWS_ACCESS_KEY_ID-}" | \
   exit 1
 fi
 
-hack/ci/e2e-conformance.sh --verbose
+# TODO(dims): remove "SKIP_INIT_IMAGE" once we fix problem building images in CI
```
Comment (aojea) [inline scripts/ci-conformance.sh]: The images are built using kubernetes master?

File: hack/heartbeat_account.py (hunk 3 above)
Comment (dims) [reply hack/heartbeat_account.py]: don't want this to be resilient, if things fail we should bail out

File: scripts/ci-conformance.sh (hunk 4 above)
Comment (dims) [reply scripts/ci-conformance.sh]: by default we pick up images with k8s 1.16.2

File: hack/ci/e2e-conformance.sh [hunk 5]
```diff
@@ -174,8 +181,26 @@ generate_manifests() {
...
+      --region ${AWS_REGION} --output json | jq -r '.[0][0] | select (.!=null)')
+    if [[ -z "$image_id" ]]; then
+      echo "unable to find image using : $filter $IMAGE_LOOKUP_ORG ... bailing out!"
+      exit 1
+    fi
+  else
+    image_id=$(aws ec2 describe-images --image-ids $IMAGE_ID \
+      --query 'Images[*].[ImageId,Name]' --output json | jq -r '.[0][0] | select (.!=null)')
+    echo "using specified image id : ${IMAGE_ID}"
+    if [[ -z "$image_id" ]]; then
```
Comment (aojea) [inline hack/ci/e2e-conformance.sh]: nit, this block seems redundant and can be used outside the main if block
```
if [[ -z ${IMAGE_ID:-} ]];
  ...
else
  ...
fi
if [[ -z "$image_id" ]]; then
  echo "unable to find image using : $filter $IMAGE_LOOKUP_ORG ... bailing out!"
  exit 1
fi
```

**Comment by aojea** [comment]
this is an impressive work.
I'd be interested in adapting it to run conformance against an IPv6 cluster, is it something feasible?

File: examples/machinedeployment/machinedeployment.yaml (hunk 2 above)
Comment (aojea) [reply examples/machinedeployment/machinedeployment.yaml]: is there a load balancer in front of the controllers for the APIs?

**Comment by aojea** [comment]
las run failed because the process of provisioning the workers got stuck
```
1030 02:25:31.968] test1-controlplane-0          aws:////i-0d55666950be00782   running
I1030 02:25:31.969] test1-controlplane-1          aws:////i-09b6460ab6b13bf08   running
I1030 02:25:31.969] test1-controlplane-2          aws:////i-0b79b4553c5f26669   running
I1030 02:25:31.969] test1-md-0-6c94dcb947-j4jxg                                 provisioning
I1030 02:25:31.969] test1-md-0-6c94dcb947-jj65z                                 provisioning
I1030 02:25:31.970] cluster did not start in 30 mins ... bailing out!
I1030 02:25:31.970] === versions ===
```
/test pull-cluster-api-provider-aws-conformance

**Comment by detiber** [comment]
@aojea We've discovered what we believe is an issue with the aws janitor that is being run by boskos, where it is not fully cleaning the account prior to it being issued to a new job. We believe this is the reason why the workers are failing to provision at this time.

**Comment by dims** [comment]
@aojea yes it's feasible once we get this ready a bit more :)

Also @detiber i added some code to force clean and then create stack, lets see if that helps

**Comment by aojea** [comment]
/test pull-cluster-api-provider-aws-conformance
progress :muscle:

**Pull request merged**